File: Code Matlab/05. Prediction/05.1 Linear Regression/Introduction.m
clc, clear, close all

dx = .1;
x = (-2:dx:2)';
y = 3*x + 3*randn(size(x))+5; 

d = [ones(length(x),1), x];

%Pseudo Inverse
[U,S,V] = svd(d,'econ');
m = V*inv(S)*U'*y;

figure('Name','Linear Regression');
hold on
scatter(x,y);
plot(x,d*m,'r');
plot(x,3*x+5,'b--');
xlabel('x')
ylabel('y')
legend('Data','Linear Regression','True Fit');

File: Code Python/05. Prediction/05.1 Linear Regression/Introduction.py
import matplotlib.pyplot as plt
import numpy as np
from functools import reduce

dx = 0.1
x = np.transpose(np.arange(-2, 2, dx))
x = x.reshape((len(x), 1))
d = np.ones(((x.shape[0]), (x.shape[1]+1)))
d[:, 1:] = x
y = 3*x + 3*np.random.randn(40, 1) + 4

# Pseudo Inverse
[U, S, V] = np.linalg.svd(d, full_matrices=False)
m = reduce(np.dot, [np.matrix.transpose(V), np.linalg.inv(np.diag(S)), np.matrix.transpose(U), y])


plt.figure()
plt.scatter(x, y)
plt.plot(x, d.dot(m), color='r')
plt.plot(x, 3*x+4, color='b', linestyle='--')
plt.xlabel('x')
plt.ylabel('y')
plt.legend(['Data', 'Linear Regression', 'True Fit'])
plt.show()

